### PR TITLE
Fix Slice shape inference crashes with SIGABRT on empty dimensions

### LINKED ISSUE #7735: [Bug] Slice shape inference crashes with SIGABRT on empty dimensions (std::clamp UB)
# Slice shape inference crashes on empty dimension with negative step (std::clamp UB)

## Summary

Slice shape inference calls `std::clamp(start, 0, dim_value - 1)` when `dim_value = 0`, producing `std::clamp(start, 0, -1)` which violates the `lo <= hi` precondition of `std::clamp`. This is undefined behavior per the C++ standard.

Since ONNX 1.21.0 enables `_GLIBCXX_ASSERTIONS` (PR #7601, OpenSSF compiler hardening), this silent UB now causes a hard crash (`SIGABRT`) on Linux with GCC/libstdc++.

## Affected Locations

The bug exists in **3 locations** across 2 files:

| File | Line | Opset | Code |
|------|------|-------|------|
| `onnx/defs/tensor/old.cc` | 2646 | Slice opset 11 | `std::clamp(start, 0, input_dim_value - 1)` |
| `onnx/defs/tensor/old.cc` | 6329 | Slice opset 10 | `std::clamp(start, 0, input_dim_value - 1)` |
| `onnx/defs/tensor/defs.cc` | 801 | Slice opset 13 (via `processSliceInputs`) | `std::clamp(start, 0, input_rank - 1)` |

Note: In `defs.cc`, the function `processSliceInputs` has a parameter named `input_rank`, but at call site line 980 it actually receives `input_dim.dim_value()`, not the tensor rank. The naming is misleading.

### Prerequisites

```bash
pip install onnx==1.21.0rc2  # or any onnx >= 1.21.0
# Must be on Linux with GCC/libstdc++ for _GLIBCXX_ASSERTIONS to be active
```

### Reproduction Script

```python
"""Reproduction: Slice shape inference std::clamp UB on empty dimension."""

import onnx
from onnx import TensorProto, helper, shape_inference

# Create a Slice model: input shape [0, 6], slice axis=0 with step=-1
X = helper.make_tensor_value_info("X", TensorProto.FLOAT, [0, 6])
Y = helper.make_tensor_value_info("Y", TensorProto.FLOAT, None)

starts = helper.make_tensor("starts", TensorProto.INT64, [1], [1])
ends = helper.make_tensor("ends", TensorProto.INT64, [1], [0])
axes = helper.make_tensor("axes", TensorProto.INT64, [1], [0])
steps = helper.make_tensor("steps", TensorProto.INT64, [1], [-1])

slice_node = helper.make_node("Slice", ["X", "starts", "ends", "axes", "steps"], ["Y"])
graph = helper.make_graph([slice_node], "repro", [X], [Y], initializer=[starts, ends, axes, steps])
model = helper.make_model(graph, opset_imports=[helper.make_opsetid("", 13)])
# Note: we skip onnx.checker.check_model() because the checker rejects outputs
# without a shape proto. The bug is in shape inference, not the model structure.

# On onnx >= 1.21.0 with _GLIBCXX_ASSERTIONS: crashes with SIGABRT here.
# On older onnx: completes with silent UB (output shape happens to be correct).
shape_inference.infer_shapes(model)
```

### Crash Output

```
/opt/rh/gcc-toolset-14/root/usr/include/c++/14/bits/stl_algo.h:3626:
constexpr const _Tp& std::clamp(const _Tp&, const _Tp&, const _Tp&)
  [with _Tp = long int]:
Assertion '!(__hi < __lo)' failed.
Aborted (SIGABRT)
```

## Expected vs Actual Behavior

**Expected**: Shape inference should handle empty dimensions correctly. Slicing an empty dimension (dim_value=0) with any step should produce an empty dimension (dim_value=0) in the output.

**Actual**: `std::clamp(start, 0, dim_value - 1)` is called with `dim_value = 0`, resulting in `std::clamp(start, 0, -1)`. Since `lo=0 > hi=-1`, this violates `std::clamp`'s precondition (`lo <= hi`), which is undefined behavior. With `_GLIBCXX_ASSERTIONS` enabled (since PR #7601), this causes a runtime assertion failure and `SIGABRT`.

## Root Cause Analysis

The Slice shape inference code assumes `dim_value >= 1` when computing clamping bounds. The pattern appears in all three locations:

```cpp
// old.cc (opset 10 and 11) — inlined in the lambda:
const auto input_dim_value = input_dim.dim_value();
// ...
if (step < 0)
  start = std::clamp(start, static_cast<int64_t>(0), input_dim_value - 1);
  //                                                  ^^^^^^^^^^^^^^^^ = -1 when dim_value=0
```

```cpp
// defs.cc (opset 13) — in processSliceInputs():
static void processSliceInputs(const int64_t input_rank, ...) {
  // "input_rank" is a misnomer — it receives dim_value at call site line 980
  if (step < 0)
    start = std::clamp(start, static_cast<int64_t>(0), input_rank - 1);
    //                                                  ^^^^^^^^^^ = -1 when 0 is passed
```

The `end` clamping has a similar issue: `std::clamp(end, -1, input_dim_value)` is fine when `input_dim_value=0` (since `-1 <= 0`), but `std::clamp(end, -1, input_dim_value - 1)` would also fail. Currently only the `step < 0` path for `start` is definitely broken, but the `end` path with `step < 0` also has `clamp(end, -1, dim_value)` which narrowly avoids the bug only because `-1 <= 0`.

PR #7601 (`b75eda1f9`) enabled `_GLIBCXX_ASSERTIONS` via OpenSSF Silver Badge compiler hardening:
```cmake
# cmake/Utils.cmake
target_compile_definitions(${target} PRIVATE _GLIBCXX_ASSERTIONS)
```

This is the correct thing to do — the hardening flag is exposing a real bug, not creating one.

## Suggested Fix

Add a `dim_value == 0` early-exit guard before the `std::clamp` calls. When a dimension is empty, the output dimension is always 0 regardless of start/end/step.

### For `old.cc` (opset 10 and 11) — add before the `std::clamp` block:

```cpp
const auto input_dim_value = input_dim.dim_value();

// Empty dimension: output is also empty, skip std::clamp
if (input_dim_value == 0) {
  ctx.getOutputType(0)
      ->mutable_tensor_type()
      ->mutable_shape()
      ->mutable_dim(static_cast<int>(axis))
      ->set_dim_value(0);
  continue;
}
```

### For `defs.cc` — add at the start of `processSliceInputs()`:

```cpp
static void processSliceInputs(const int64_t input_rank, int64_t& start, int64_t& end, int64_t step) {
  if (step == 0) {
    fail_shape_inference("'step' cannot be 0 for Slice");
  }
  // Guard: when called with dim_value=0, clamp bounds would be invalid
  if (input_rank == 0) {
    start = 0;
    end = 0;
    return;
  }
  // ... rest of function unchanged
```

### Reference: ONNX Runtime already handles this correctly

ONNX Runtime's `slice_helper.h` (line 116) has the correct pattern:

```cpp
if (dim_value == 0) {
  // shape with empty dim. only output_dims_ matters but set everything for completeness
  compute_metadata.steps_[axis] = step;
  compute_metadata.starts_[axis] = 0;
  compute_metadata.ends_[axis] = 0;
  compute_metadata.output_dims_[axis] = 0;
  continue;
}
```

## Impact

- **Crashes all downstream consumers** that run ONNX shape inference on models with empty dimensions and negative-step slicing
- **ONNX Runtime CI is affected**: `SliceTest.EmptyDim` crashes on all Release builds
- **Severity**: High — this is a regression in 1.21.0; models that previously worked now crash
- **Scope**: Any ONNX model with `dim_value=0` + `Slice(step=-1)` on that axis

## Repro

```python
"""
Reproduction script for ONNX Slice shape inference std::clamp UB bug.

Bug: When a tensor has an empty dimension (dim_value=0) and the slice step is
negative, ONNX shape inference calls:
    std::clamp(start, 0, dim_value - 1)  →  std::clamp(start, 0, -1)
This violates the std::clamp precondition (lo <= hi), which is undefined behavior.

ONNX 1.21.0 enables _GLIBCXX_ASSERTIONS (PR #7601, OpenSSF hardening), which
turns this silent UB into a runtime assertion failure (SIGABRT).

Affected locations (all in onnx/defs/tensor/):
  - old.cc:2646  — Slice opset 11 shape inference
  - old.cc:6329  — Slice opset 10 shape inference
  - defs.cc:792  — processSliceInputs() used by Slice opset 13

Requirements:
  pip install onnx==1.21.0rc2  (or any onnx >= 1.21.0 built with _GLIBCXX_ASSERTIONS)
  Linux with GCC / libstdc++

Expected result on onnx >= 1.21.0 (with _GLIBCXX_ASSERTIONS):
  Process crashes with SIGABRT:
    std::clamp(const _Tp&, const _Tp&, const _Tp&): Assertion '!(__hi < __lo)' failed.

On older onnx versions (< 1.21.0) without _GLIBCXX_ASSERTIONS:
  Shape inference completes (silent UB) — output happens to be correct by accident.
"""

import onnx
from onnx import TensorProto, helper, shape_inference


def make_slice_model(opset_version: int) -> onnx.ModelProto:
    """Create a minimal Slice model with an empty dimension and negative step.

    The model slices a tensor of shape [0, 6] on axis 0 (the empty dimension)
    with step=-1. This triggers std::clamp(start, 0, dim_value-1) where
    dim_value=0, causing clamp(start, 0, -1) — undefined behavior.
    """
    # Input: shape [0, 6] — first dimension is empty
    X = helper.make_tensor_value_info("X", TensorProto.FLOAT, [0, 6])

    # Slice parameters as constant initializers
    starts = helper.make_tensor("starts", TensorProto.INT64, [1], [1])
    ends = helper.make_tensor("ends", TensorProto.INT64, [1], [0])
    axes = helper.make_tensor("axes", TensorProto.INT64, [1], [0])
    steps = helper.make_tensor("steps", TensorProto.INT64, [1], [-1])

    # Output: leave shape unspecified so shape inference must compute it.
    # We intentionally do NOT call onnx.checker.check_model() here because the
    # checker rejects outputs without a shape proto. The bug is in shape
    # inference (infer_shapes), not in the model structure itself.
    Y = helper.make_tensor_value_info("Y", TensorProto.FLOAT, None)

    slice_node = helper.make_node(
        "Slice",
        inputs=["X", "starts", "ends", "axes", "steps"],
        outputs=["Y"],
    )

    graph = helper.make_graph(
        [slice_node],
        f"slice_empty_dim_opset{opset_version}",
        [X],
        [Y],
        initializer=[starts, ends, axes, steps],
    )

    model = helper.make_model(graph, opset_imports=[helper.make_opsetid("", opset_version)])
    return model


def test_opset(opset_version: int) -> None:
    """Test shape inference for a given opset version.

    Opset 10 and 11 use old.cc; opset 13 uses defs.cc. All three have the bug.

    On ONNX >= 1.21.0 with _GLIBCXX_ASSERTIONS enabled, this will crash the
    process with SIGABRT (not a Python exception — a hard C++ abort).
    """
    print(f"\n--- Testing Slice shape inference with opset {opset_version} ---")
    model = make_slice_model(opset_version)
    print(f"  Model created: input shape [0, 6], slice axis=0, step=-1")
    print(f"  Running onnx.shape_inference.infer_shapes()...")
    print(f"  (If _GLIBCXX_ASSERTIONS is enabled, this will SIGABRT here)")

    try:
        inferred = shape_inference.infer_shapes(model)
        out_shape = inferred.graph.output[0].type.tensor_type.shape
        dims = [d.dim_value for d in out_shape.dim]
        print(f"  Completed — inferred output shape: {dims}")
        if dims == [0, 6]:
            print(f"  Shape is correct (but std::clamp UB was silently triggered)")
        else:
            print(f"  ERROR: Expected [0, 6], got {dims} — UB produced wrong result")
    except Exception as e:
        print(f"  Exception: {type(e).__name__}: {e}")


if __name__ == "__main__":
    print(f"ONNX version: {onnx.__version__}")
    print()
    print("Bug: std::clamp(start, 0, dim_value-1) when dim_value=0")
    print("     → std::clamp(start, 0, -1) violates lo<=hi precondition")
    print()
    print("On onnx >= 1.21.0 with _GLIBCXX_ASSERTIONS (Linux/GCC):")
    print("  → SIGABRT crash (assertion catches the UB)")
    print("On onnx < 1.21.0 or without _GLIBCXX_ASSERTIONS:")
    print("  → Silent UB (shape inference may complete with correct or wrong result)")

    # Test all affected opset versions.
    # If _GLIBCXX_ASSERTIONS is active, the first opset that hits the bug
    # will crash the process with SIGABRT (no Python exception — hard crash).
    for opset in [10, 11, 13]:
        test_opset(opset)

    print("\n✅ All tests completed without crash.")
    print("   (Either the bug is fixed, or _GLIBCXX_ASSERTIONS is not enabled.)")

```

### PR DESCRIPTION
**Pull request by titaiwangms**
Fix #7735 

This pull request addresses issues with the ONNX Slice operator when handling empty dimensions, ensuring correct and safe shape inference in these edge cases. The main focus is to prevent undefined behavior and ensure that slicing on empty dimensions consistently produces outputs with zero length, regardless of the step direction. Additional regression tests are included to verify these fixes, and several test data files are updated to reflect the new behavior.

**Slice operator improvements for empty dimensions:**

* Updated the `processSliceInputs` function in `defs.cc` to short-circuit and produce a zero-length output when the input dimension size is zero, avoiding invalid clamp bounds and potential undefined behavior.
* Modified shape inference logic in `old.cc` to detect empty dimensions and set the output dimension value to zero, both for standard and expanded shape inference code paths. [[1]](diffhunk://#diff-5e200991755406044e2dbb2449683bb5c6d77583d1459f0065a756813dd1b59bR2635-R2645) [[2]](diffhunk://#diff-5e200991755406044e2dbb2449683bb5c6d77583d1459f0065a756813dd1b59bR6329-R6335)

**Testing and regression coverage:**

* Added two new tests in `shape_inference_test.py` to verify that slicing on empty dimensions with both positive and negative steps produces the correct output shape, preventing regression of issue #7735.

**Test data updates:**

* Updated several `output_0.pb` binary test data files under `onnx/backend/test/data/node/` (including attention, gelu, hammingwindow, and lppool tests) to align with the corrected Slice operator behavior on empty dimensions. [[1]](diffhunk://#diff-875055c843db41ea7d579c8bd29201e6f60bb4175ac96f840bdd27d488866198L1-R7) [[2]](diffhunk://#diff-d6fe5631baa8b6cb320c03143f797abe3dce4ddf5363fa9cfe71cc12bc31ac53L2-R3) [[3]](diffhunk://#diff-3275d5e3517a0e95967de3c46a75b02104980f2566f1a0cbefa386ea8e02f312L1-R3) [[4]](diffhunk://#diff-537fe44987ef880d522c9c88283070745874abc990951f49eff8dccc35630af1L2-R2) [[5]](diffhunk://#diff-3d57e2f0cf63b32349f6619a2f61e90d4a4aa4fdac9a794af6bfb00c97fe9f8eL1-R2)

These changes collectively improve the robustness of ONNX shape inference and operator correctness when handling tensors with empty dimensions.

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -794,6 +794,13 @@ static void processSliceInputs(const int64_t input_rank, int64_t& start, int64_t
   if (step == 0) {
     fail_shape_inference("'step' cannot be 0 for Slice");
   }
+  // Empty dimension: clamp bounds are invalid when dimension size is 0,
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:797]: I don't see a test-case for input_rank == 0 (in the new tests)

Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:797]: The above comment in the code should say "input rank" not "dimension size"

Comment (titaiwangms) [reply onnx/defs/tensor/defs.cc:797]: So there are two callsites to this function. One is using dim_size, while the other one is using dim_value. This is just bad naming. The PR is fixing dim_value==0.

Comment (titaiwangms) [reply onnx/defs/tensor/defs.cc:797]: See above reply

Comment (titaiwangms) [reply onnx/defs/tensor/defs.cc:797]: I will do a follow-up.

Comment (titaiwangms) [reply onnx/defs/tensor/defs.cc:797]: https://github.com/onnx/onnx/pull/7739
